### ISSUE TITLE
Listing of commands

### ISSUE THREAD

**Issue by 4C3T** [luadch/luadch#122]
Hi. I'm brand new to this hubsoft, got it installed in a jiffy, no problems at all (Linux Mint x64).
Everything's just fine, but I'm super annoyed about the !help function, or more precisely what it lists.

For eg (Last few lines):
"etc_records.lua - Owners
Usage:		[+!#]records show|reset
Description:	Sends the hub records to user|reset records database
Min. Level:	100"

When read like this it's ok, but bundled with all the rest, and much smaller fonts makes it very cluttered and hard to read, so I'm wondering which file is responsible?
Had a quick look in the folder /scripts/cmd_help.lua and found a row that might be what I'm looking for, but I'm not sure.
I would love to have a simple clean listing of the commands, and preferably listed by group, but that's perhaps a bit too much to hope for I suppose.

Anyone who knows which file is handling this, and/or a quick fix?

**Comment by blasti** [comment]
> so I'm wondering which file is responsible?

There is no single responsible file. Most commands are defined in their own script.
For example, the command "ban" is defined in "scripts/cmd_ban.lua"
At hubstart, each script can register a help description of its via the script "cmd_help.lua".

Later, if some user executes "help", the "cmd_help.lua" script assembles all descriptions according to
the level of the user. That is, if a command has a higher level than the user, its description won't be printed.

If you, as a hub admin, execute "help", you will get a long list of every command that exists.
Try executing "help" as an low level user, and you'll have your "simple, clean" listing.

Hope that helps.

**Comment by 4C3T** [comment]
Ok. Not much of a way to fix it.
Too bad, I was hoping to at least compact it a bit so it won't look so cluttered. It's time consuming when you're trying to find a specific command. This is quite a mess having to sift through

**Comment by 4C3T** [comment]
It'd be one thing if there could be two !help commands, one brief and more extensive. That'd be more useful, and above else, much quicker too.

**Comment by blasti** [comment]
> It'd be one thing if there could be two !help commands, one brief and more extensive.

This could be done, if one knew what how you define "brief" in this context.

**Comment by 4C3T** [comment]
Well, just a list of commands without unnecessary characters like [!/+/#].
For eg:
"etc_trafficmanager.lua - Owners
Usage:		[+!#]trafficmanager block <NICK> [<REASON>] | unblock <NICK>
Description:	Blocks downloads ( d ), uploads ( u ) and search ( s ) | Unblock user
Min. Level:	60"

could be shorted down to:

"trafficmanager block <NICK> [<REASON>] | unblock <NICK> -Blocks D/L, U/L and Search. Min. LVL 60" One per row would be nice...

The command characters "+!#" are obvious and can be noted at the top, as well as syntax examples.
It's too messy with, what, 30-40 sets of these... Especially if you're in a hurry or stressed to find that specific command.

**Comment by blasti** [comment]
Feel free to change the formatting according to your needs. It is determined in the section following this line:

https://github.com/luadch/luadch/blob/master/scripts/cmd_help.lua#L84

You can also commit a pull request for a "!help brief" extension.

**Comment by 4C3T** [comment]
Yeah, I found that and thought it looked like I was on the right track, but it was something that didn't quite look right, whatever it was that I was thinking of.
I'll have a look at it again.

I'm not really a programmer, but I've done some tinkering in my days.
\t means tab, and \n new line right? I think it was one of those that didn't quite match what I saw in that listing.. Oh, well... I'll have a look at it, as I said.

**Comment by 4C3T** [comment]
I did manage to slim it down, but damn it's hard to edit it, slightest little wrong and the hub won't recognize !help as a command.
I find it strange that I can't just remove certain strings.

```
tbl.title .. "\n" ..
                              msg_usage .. "\t\t" .. tbl.usage .. "\n" ..
                              msg_description .. "\t" .. tbl.desc .. "\n" ..
                              msg_minlevel .. "\t" .. tbl.level .. "\n"
```

What I got so far:
```
tbl.title .. "\n" ..
                             tbl.usage .. "\t" ..
                             msg_description .. tbl.desc ..
                             msg_minlevel .. tbl.level .. "\t"

```
It does work, but I haven't been able to find what places the annoying tags "[+!#]" in front of every listed command
It really makes it look messy

As for the listing for the command "offline" I get this:
"- [+!#]offline [ show | addexception | removeexceptions | showexceptions | delete [a, b, c, ...] | showdeleted | autoclean | showsettings | set [ toggleautoclean | maxlvl <VALUE> | days <VALUE> ] | help]	Description:Allows you to [ shows all users who have been offline for too long | add an exception | remove an exception | delete users by number | shows all by this script deleted users | start autoclean | show settings | set [ toggle auto clean schedule | max level to check (manual clean) | max offline days (manual clean)] ].Min. Level:80	
"

This takes up two rows of text and I'd really like to slim that down a bit, so the question is where do I find this comment, or whatever to call it?

Also, when I last looked at the "!errors" I got a heap of errors pertaining to the help-command, and I didn't think to run this prior to changing the lua-file.
That says:

```
- [2020-12-04 | 17:42:49] server.lua: function 'wrapserver': wrong sslctx parameters: error loading private key (No such file or directory)
- [2020-12-04 | 17:42:49] server.lua: function 'wrapserver': wrong sslctx parameters: error loading private key (system lib)
- [2020-12-05 | 18:17:55] server.lua: function 'addserver', luasocket socket bind: address already in use
- [2020-12-05 | 18:17:55] server.lua: function 'addserver', luasocket socket bind: address already in use
- [2020-12-06 | 17:11:57] scripts.lua: syntax error in script 'cmd_help.lua': ././scripts/cmd_help.lua:89: unexpected symbol near 'end'
- [2020-12-06 | 17:29:26] scripts.lua: syntax error in script 'cmd_help.lua': ././scripts/cmd_help.lua:89: unexpected symbol near 'end'
- [2020-12-06 | 17:34:46] scripts.lua: syntax error in script 'cmd_help.lua': ././scripts/cmd_help.lua:90: unexpected symbol near 'end'
- [2020-12-06 | 17:35:28] scripts.lua: syntax error in script 'cmd_help.lua': ././scripts/cmd_help.lua:90: unexpected symbol near 'end'
- [2020-12-06 | 17:40:32] scripts.lua: syntax error in script 'cmd_help.lua': ././scripts/cmd_help.lua:85: unexpected symbol near '..'
- [2020-12-06 | 17:42:31] scripts.lua: syntax error in script 'cmd_help.lua': ././scripts/cmd_help.lua:87: unexpected symbol near 'end'
- [2020-12-06 | 17:42:56] scripts.lua: syntax error in script 'cmd_help.lua': ././scripts/cmd_help.lua:86: unexpected symbol near '..'
- [2020-12-06 | 17:43:13] scripts.lua: syntax error in script 'cmd_help.lua': ././scripts/cmd_help.lua:86: '=' expected near '..'
- [2020-12-06 | 17:43:42] scripts.lua: syntax error in script 'cmd_help.lua': ././scripts/cmd_help.lua:87: unexpected symbol near 'end'
- [2020-12-06 | 17:44:03] scripts.lua: syntax error in script 'cmd_help.lua': ././scripts/cmd_help.lua:88: unexpected symbol near 'end'
- [2020-12-06 | 17:45:35] scripts.lua: syntax error in script 'cmd_help.lua': ././scripts/cmd_help.lua:87: unexpected symbol near 'end'
- [2020-12-06 | 17:50:30] scripts.lua: syntax error in script 'cmd_help.lua': ././scripts/cmd_help.lua:86: unexpected symbol near '..'
- [2020-12-06 | 17:54:29] scripts.lua: script error: ././scripts/cmd_help.lua:99: attempt to read undeclared var: 'ucmd' (listener: onStart; script: 'cmd_help.lua')
- [2020-12-06 | 17:55:21] scripts.lua: script error: ././scripts/cmd_help.lua:99: attempt to read undeclared var: 'ucmd' (listener: onStart; script: 'cmd_help.lua')
- [2020-12-06 | 18:30:04] server.lua: function 'addserver', luasocket socket bind: address already in use
- [2020-12-06 | 18:30:04] server.lua: function 'addserver', luasocket socket bind: address already in use
- [2020-12-06 | 18:36:25] server.lua: function 'addserver', luasocket socket bind: address already in use
- [2020-12-06 | 18:36:25] server.lua: function 'addserver', luasocket socket bind: address already in use


```

**Comment by blasti** [comment]
> I did manage to slim it down, but damn it's hard to edit it, slightest little wrong and the hub won't recognize !help as a command

That's programming.

> It does work, but I haven't been able to find what places the annoying tags "[+!#]" in front of every listed command

The tags are defined in each script. See e.g. https://github.com/luadch/luadch/blob/master/scripts/cmd_rules.lua#L60
You could strip it away with [string.gsub:](https://www.lua.org/manual/5.1/manual.html#pdf-string.gsub)

```
local usage_stripped = string.gsub( tbl.usage, "[%[%]#%+!]", "" )

tbl.title .. "\n" ..

                             usage_stripped .. "\t" ..
                             msg_description .. tbl.desc ..
                             msg_minlevel .. tbl.level .. "\t"
```

> This takes up two rows of text and I'd really like to slim that down a bit, so the question is where do I find this comment, or whatever to call it?

The description of the command is stored in `tbl.desc`

> [...]  I got a heap of errors 

Your error log says basically: SSL doesn't work, probably because you didn't create a certificate. Also some socket is already in use. The script errors are because you did something wrong in the cmd_help.lua script.

**Comment by 4C3T** [comment]
I'll have a look at it at my earliest convenience.
I did however fix that other problem with SSL, except for the IPv6 since my current VPN ISP doesn't support it at my current configuration.
Couldn't run the .sh script since something was probably missing, so I did it manually as instructed in the file. Took a few tries before it worked.

It could be noted that the <> tag for the keyprint has to be included in the cfg.tbl to make it crystal clear.
I'm not really a Linux hard core user, and I believe many others aren't either, so it would be nice to have some info at n0ob level.
There could a an example included right next to the annotation.
